SYCL/ESIMD/api/functional/ctors/common.hpp

```diff
@@ -0,0 +1,165 @@
+    }
+    return min_val;
+  }
+
+  static DataT nan() {
+    DataT nan_val{};
+    if constexpr (std::is_same_v<DataT, double>) {
+      nan_val = sycl::nan(42UL);
+    } else if constexpr (std::is_same_v<DataT, float>) {
+      nan_val = sycl::nan(42U);
+    } else if constexpr (std::is_same_v<DataT, sycl::half>) {
+      nan_val = 0;
+    }
+    return nan_val;
+  }
+
+  static DataT max() {
+    DataT max_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      max_val = 0;
+    } else {
+      max_val = std::numeric_limits<DataT>::max();
+    }
+    return max_val;
+  }
+};
+
+// Generate std::vector with reference according to current tested data type and
+// number of elements.
+template <typename DataT, int NumElems>
+void generate_ref_data(vector_with_allocator<DataT> &ref_data) {
+
+  const DataT min = value<DataT>::lowest();
+  const DataT min_half = min / 2;
+  const DataT max = value<DataT>::max();
+  const DataT max_half = max / 2;
+  const DataT inf = value<DataT>::inf();
+  const DataT min_plus_one = min + 1;
+  const DataT max_minus_one = max - 1;
+
+  if constexpr (std::is_signed_v<DataT>) {
+    if constexpr (NumElems == 1) {
+
+      ref_data.insert(ref_data.end(), {min, min_half, max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {min, min_plus_one, min_half, max,
+                                       max_minus_one, max_half, -1, 0});
+      for (size_t it = 8; it < NumElems; it++) {
+        ref_data.push_back(it);
+      }
+    }
+  }
+
+  else if constexpr (std::is_unsigned_v<DataT>) {
+    if constexpr (NumElems == 1) {
+      ref_data.insert(ref_data.end(), {max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {max, max_minus_one, max_half, 0});
+      for (size_t it = 4; it < NumElems; it++) {
+        ref_data.push_back(it);
+      }
+    }
+  }
+
+  else if constexpr (std::is_same_v<DataT, sycl::half> ||
+                     std::is_floating_point_v<DataT>) {
+    DataT nan{value<DataT>::nan()};
+
+    if constexpr (NumElems == 1) {
+      ref_data.insert(ref_data.end(), {-inf, nan, min, max, -0.0, 0.1});
+    } else {
+      ref_data.insert(ref_data.end(),
+                      {-inf, nan, min, max, -0.1, -0.0, +0.0, +0.1});
+      for (size_t it = 8; it < NumElems; it++) {
+        ref_data.push_back(it + 0.25);
+      }
+    } // else
+  }
+}
+
+// Calling simd constructor in provided invocation context, it depends on the
+// TestCaseT type. TestCaseT is a struct, that should have call_simd_ctor method
+// that should return constructed object of simd class.
+// This function returning std::vector with output data.
+template <typename DataT, int NumElems, typename TestCaseT>
+auto call_simd(sycl::queue &queue,
+               const vector_with_allocator<DataT> &ref_data) {
+
+  vector_with_allocator<DataT> result{NumElems, allocator_type<DataT>{queue}};
+
+  queue.submit([&](sycl::handler &cgh) {
+    auto ref = ref_data.data();
```

Comment:
```suggestion
    const auto ref = ref_data.data();
```
Is it possible to enforce const-correctness for the reference data?

Reply:
I've done it in this commit: 0bab576359c48f2c44e6362dd3a7cb18d517def8